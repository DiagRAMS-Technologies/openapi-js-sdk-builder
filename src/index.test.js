import { generateSDKFromOpenAPI } from './index';

describe('generateSDKFromOpenAPI', () => {
  test('should work', async () => {
    expect(
      await generateSDKFromOpenAPI(
        JSON.stringify({
          openapi: '3.0.2',
          info: {
            version: '3.1.3',
            title: '@whook/example',
            description: 'A basic Whook server',
          },
          servers: [{ url: 'http://192.168.10.149:8000/v3' }],
          components: {
            securitySchemes: {
              bearerAuth: {
                description: 'Bearer authentication with a user API token',
                type: 'http',
                scheme: 'bearer',
              },
              fakeAuth: {
                description: 'A fake authentication for development purpose.',
                type: 'apiKey',
                in: 'header',
                name: 'Authorization',
              },
            },
          },
          paths: {
            '/openAPI': {
              get: {
                operationId: 'getOpenAPI',
                summary: 'Get API documentation.',
                tags: ['system'],
                responses: {
                  '200': {
                    description: 'Provides the private Open API documentation',
                    content: {
                      'application/json': { schema: { type: 'object' } },
                    },
                  },
                },
                security: [
                  {},
                  { bearerAuth: ['admin'] },
                  { fakeAuth: ['admin'] },
                ],
                parameters: [],
              },
            },
            '/ping': {
              get: {
                operationId: 'getPing',
                summary: "Checks API's availability.",
                tags: ['system'],
                responses: {
                  '200': {
                    description: 'Pong',
                    content: {
                      'application/json': {
                        schema: {
                          type: 'object',
                          additionalProperties: false,
                          properties: {
                            pong: { type: 'string', enum: ['pong'] },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
            '/delay': {
              get: {
                operationId: 'getDelay',
                summary: 'Answer after a given delay.',
                tags: ['system'],
                parameters: [
                  {
                    in: 'query',
                    name: 'duration',
                    required: true,
                    description: 'Duration in milliseconds',
                    schema: { type: 'number' },
                  },
                ],
                responses: { '204': { description: 'Delay expired' } },
              },
            },
            '/diag': {
              get: {
                operationId: 'getDiagnostic',
                summary: "Returns current API's transactions.",
                security: [{ bearerAuth: ['admin'] }, { fakeAuth: ['admin'] }],
                tags: ['system'],
                parameters: [],
                responses: {
                  '200': {
                    description: 'Diagnostic',
                    content: {
                      'application/json': {
                        schema: { type: 'object', additionalProperties: true },
                      },
                    },
                  },
                },
              },
            },
            '/time': {
              get: {
                operationId: 'getTime',
                summary: 'Get API internal clock date.',
                tags: ['system'],
                responses: {
                  '200': {
                    description: 'Server current date',
                    content: {
                      'application/json': {
                        schema: {
                          type: 'object',
                          additionalProperties: false,
                          properties: {
                            currentDate: {
                              type: 'string',
                              format: 'date-time',
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
            '/echo': {
              put: {
                operationId: 'putEcho',
                summary: 'Echoes what it takes.',
                tags: ['system'],
                requestBody: {
                  description: 'The input sentence',
                  content: {
                    'application/json': {
                      schema: {
                        type: 'object',
                        required: ['echo'],
                        additionalProperties: false,
                        properties: { echo: { type: 'string' } },
                      },
                      example: { echo: 'Repeat this!' },
                    },
                  },
                },
                responses: {
                  '200': {
                    description: 'The actual echo',
                    content: {
                      'application/json': {
                        schema: {
                          type: 'object',
                          required: ['echo'],
                          additionalProperties: false,
                          properties: { echo: { type: 'string' } },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
          tags: [],
        }),
      ),
    ).toMatchSnapshot();
  });
});
