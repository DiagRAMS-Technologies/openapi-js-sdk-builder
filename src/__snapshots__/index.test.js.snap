// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSDKFromOpenAPI should work 1`] = `
"
// WARNING: This file is automatically generated
// do not change it in place it would be overridden
// by the next build
'use strict';

const querystring = require('querystring');
const axios = require('axios');

/**
 * A basic Whook server
 * @module API
 * @version 3.1.3
 */
const API = {
  getOpenAPI,
  getPing,
  getDelay,
  getDiagnostic,
  getTime,
  putEcho,
};


/**
 * Get API documentation.
 * @param {Object} parameters
 * The parameters to provide (destructured)
 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function getOpenAPI(_, options) {



  const method = 'get';
  let urlParts = [
    'openAPI',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
  });
  let data = {}.undef;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

/**
 * Checks API's availability.
 * @param {Object} parameters
 * The parameters to provide (destructured)
 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function getPing(_, options) {



  const method = 'get';
  let urlParts = [
    'ping',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
  });
  let data = {}.undef;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

/**
 * Answer after a given delay.
 * @param {Object} parameters
 * The parameters to provide (destructured)
 * @param {number} parameters.duration
 * Duration in milliseconds
 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function getDelay({
  duration,} = {}, options) {


  if( duration == null) {
    throw new Error('Missing required parameter : duration. Value : ' +  duration);
  }


  const method = 'get';
  let urlParts = [
    'delay',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
    duration: duration,
  });
  let data = {}.undef;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

/**
 * Returns current API's transactions.
 * @param {Object} parameters
 * The parameters to provide (destructured)
 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function getDiagnostic(_, options) {



  const method = 'get';
  let urlParts = [
    'diag',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
  });
  let data = {}.undef;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

/**
 * Get API internal clock date.
 * @param {Object} parameters
 * The parameters to provide (destructured)
 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function getTime(_, options) {



  const method = 'get';
  let urlParts = [
    'time',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
  });
  let data = {}.undef;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

/**
 * Echoes what it takes.
 * @param {Object} parameters
 * The parameters to provide (destructured)
  @param body The request body

 * @param {Object} options
 * Options to override Axios request configuration
 * @return {Object}
 * The HTTP response
 */
function putEcho({
      body,} = {}, options) {



  const method = 'put';
  let urlParts = [
    'echo',
  ];
  let headers = Object.assign(((options || {}).headers || {}), {
    'X-API-Version': '3.1.3',
  });
  let qs = cleanQuery({
  });
  let data = body;

  return axios(Object.assign({
    baseURL: '[object Object]',
    paramsSerializer: querystring.stringify.bind(querystring),
    validateStatus: status => 200 <= status && 300 > status,
    method: method,
    url: urlParts.join('/'),
    headers: cleanHeaders(headers),
    params: qs,
    data,
  }, options || {}));
}

function cleanQuery(query) {
    return Object.keys(query)
    .filter(key => \\"undefined\\" !== typeof query[key])
    .filter(key => !(query[key] instanceof Array) || 0 !== query[key].length)
    .reduce((newQuery, key) => {
        newQuery[key] = query[key];
        return newQuery;
    }, {});
}

function cleanHeaders(headers) {
    return Object.keys(headers)
    .filter(key => \\"undefined\\" !== typeof headers[key])
    .reduce((newHeaders, key) => {
        newHeaders[key] = headers[key];
        return newHeaders;
    }, {});
}

// eslint-disable-next-line no-unused-vars
function sortMultipleQuery(a, b) {
    return a > b ? 1 : -1;
}

module.exports = API;
"
`;
